core/dataset.cpp

```diff
@@ -680,4 +744,192 @@ std::ostream &operator<<(std::ostream &os, const Dim dim) {
   return os << to_string(dim);
 }
 
+Dataset operator+(const Dataset &lhs, const Dataset &rhs) {
+  Dataset res(lhs);
+  apply(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator+(const Dataset &lhs, const DatasetConstProxy &rhs) {
+  Dataset res(lhs);
+  apply(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator+(const Dataset &lhs, const DataConstProxy &rhs) {
+  Dataset res(lhs);
+  apply_with_delay(plus_equals, res, rhs);
+  return res;
+}
+
+Dataset operator+(Dataset &&lhs, const Dataset &rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(Dataset &&lhs, const DataConstProxy &rhs) {
+  return apply_with_delay(plus_equals, lhs, rhs);
+}
+
+Dataset operator+(const Dataset &lhs, Dataset &&rhs) {
```

Comment:
I would expect `operator-` had an overload like this too.

Reply:
I don't think there is an easy way to implement this based on what operations are already available.

Reply:
I think based on non-commutative behaviour, this cannot be implemented, as you say.